# Lab 06 - Data Preprocessing I

Name: Jake Schaeffer  
Class: CSCI 349 - Intro to Data Mining  
Semester: Spring 2020  
Instructor: Brian King


In [75]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### 1) [P] Use pandas to read in your CSV data file you downloaded above, which you should have placed in yourdata directory. Call the data frame df_temps. Read in the entire dataset, however, peek at the dataset first. You'll notice 16 rows of metadata. Ignore the first 16 rows (HINT: Use the skiprows= option!)
NOTE: BE SURE TO LOOK AT YOUR ACTUAL DATA BEFORE TRYING TO READ IN A RAW DATASET! JUST BECAUSE
A DATASET HAS A .CSV EXTENSION DOES NOT MEAN THAT YOU CAN RELY ON EVERY ROW BEING A PROPERLY
FORMATTED ROW! For instance, notice that the header row is scattered throughout your data! Notice that
you have some extra columns at the end that are consistently empty! The inexperienced are tempted to
manually edit the file to make it easy to read. NO. WRONG! BAD DATA SCIENTIST! Write your Python cleaning
code to always work with raw, uncleaned data. Why? In practice, your data file may be huge. You may need to
repeatedly grab fresh data, that will only have the same issues. Do you really want to repeat your manual
editing silliness every time you have a fresh file? No! It may take a bit more work up front, but ALWAYS strive
to write code to preprocess every aspect of your data file! It will always save you work later!

In [76]:
df_temps = pd.read_csv("../data/faa_hourly-KIPT_20000101-20191231_raw.csv",skiprows = 16)
df_temps.head()

Date/Time (GMT) Number of Observations (n/a) Average Temp (F)  \
0  2000-01-01 00:00:00                            1            26.10   
1  2000-01-01 01:00:00                            1            26.10   
2  2000-01-01 02:00:00                            1            26.10   
3  2000-01-01 03:00:00                            1            26.10   
4  2000-01-01 04:00:00                            1            26.10   

  Max Temp (F) Min Temp (F) Average Dewpoint Temp (F) 1 Hour Precip (in)  \
0        26.10        26.10                     14.00                NaN   
1        26.10        26.10                     14.00                NaN   
2        26.10        26.10                     15.10                NaN   
3        26.10        26.10                     12.00                NaN   
4        26.10        26.10                     14.00                NaN   

  Max Wind Gust (mph) Average Relative Humidity (%) Average Wind Speed (mph)  \
0               20.70                         59.00                    17.30   
1                 NaN                         59.00                    16.10   
2                 NaN                         62.00                    15.00   
3                 NaN                         54.00                    16.10   
4                 NaN                         59.00                    12.70   

  Average Station Pressure (mb) Average Wind Direction (deg)  \
0                        1015.2                       280.00   
1                        1015.9                       280.00   
2                        1016.3                       280.00   
3                        1016.6                       270.00   
4                        1017.3                       280.00   

  Max Wind Speed (mph)  Unnamed: 13  
0                17.30          NaN  
1                16.10          NaN  
2                15.00          NaN  
3                16.10          NaN  
4                12.70          NaN

### 2) Report the general structure of the data frame using df_temps.info(). 
You should notice that almost
every variable was read in as a plan object data type. You have a lot of work to do!
Your result should look as follows:

This is a pretty good dataset with lots of real problems! It gives you a chance to understand how important it is to
select the smallest, yet most accurate data type for every variable. This is particularly true with respect to your
memory footprint. With enormous data involving millions of records, you often need to perform various paging
exercises to load in chunks of data into memory, substantially slowing down the machine learning methods. In
other words, the more data you can fit in memory, the better!

In [77]:
df_temps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181914 entries, 0 to 181913
Data columns (total 14 columns):
Date/Time (GMT)                  181914 non-null object
Number of Observations (n/a)     181914 non-null object
Average Temp (F)                 180915 non-null object
Max Temp (F)                     180915 non-null object
Min Temp (F)                     180915 non-null object
Average Dewpoint Temp (F)        180809 non-null object
1 Hour Precip (in)               37596 non-null object
Max Wind Gust (mph)              32206 non-null object
Average Relative Humidity (%)    177416 non-null object
Average Wind Speed (mph)         181372 non-null object
Average Station Pressure (mb)    181636 non-null object
Average Wind Direction (deg)     149252 non-null object
Max Wind Speed (mph)             181372 non-null object
Unnamed: 13                      0 non-null float64
dtypes: float64(1), object(13)
memory usage: 19.4+ MB


### 3) [P] Read about the memory_usage() method of pandas data frames. Then, report the total memory in bytes for each variable of df_temps. Set the parameter drop=True, to get the most accurate assessment of your total memory usage. (NOTE – this could take a bit of time to return an answer.)

In [78]:
df_temps.memory_usage(deep = True)

Index                                 128
Date/Time (GMT)                  13790816
Number of Observations (n/a)     11477885
Average Temp (F)                 11342670
Max Temp (F)                     11308038
Min Temp (F)                     11308017
Average Dewpoint Temp (F)        11411051
1 Hour Precip (in)                7032800
Max Wind Gust (mph)               6907372
Average Relative Humidity (%)    11354487
Average Wind Speed (mph)         11291210
Average Station Pressure (mb)    11647992
Average Wind Direction (deg)     10569351
Max Wind Speed (mph)             11259415
Unnamed: 13                       1455312
dtype: int64

### 4) [P] Report the total memory required for the data frame in MB. (Just sum the previous answer.) You should get an answer showing hundreds of megabytes! Also, store the total as a variable called original_memory. We're going to compare memory after we're done.

In [79]:
original_memory = df_temps.memory_usage(deep = True).sum() / 1000000
print("{} megabytes!!!".format(original_memory))

142.156544 megabytes!!!


### 5) [P] Remember those extra column header lines that appeared throughout the entire CSV file? You need to get rid of those. Write the code to eliminated those from df_temps. (This is tricky. Think about it... you are selecting the data that does NOT have columns[0] as the first value in the observation!)
HINT – At this point, you should have 173252 observations.

In [80]:
#select only values in df_temps where the date is not equal to the date header  - aka the rows that are not the column labels
df_temps = df_temps[df_temps[df_temps.columns[0]] != df_temps.columns[0]]
len(df_temps)

173252

### 6) [M-P] Examine the last entry of your index (i.e. df_temps.index[-1]). Then, show the number of observations in df_temps. These are unequal. Why?

In [81]:
print("last index: ",df_temps.index[-1])
print("number of observations: ",len(df_temps))

last index:  181913
number of observations:  173252


we got rid of some of the rows, lowering the number of observations. however, this did not change the index of the observations that were kept.

### 7) When you permanently delete observations, it's usually a good idea to reset your index, especially if the indexis nothing more than a unique number to each observations. Reindex your data, and show that the new index is indeed reset. 
(There are many ways to do this. I suggest using reset_index(). There is no need toretain the original index, so drop=True is fine.)

In [82]:
df_temps = df_temps.reset_index(drop = True)
df_temps.index[-1]

173251

### 8) You have a rather annoying extra column that was read in in the last column position. (Look closely at the output of `describe()` above!) You should always confirm that it's garbage before deleting it. Write the single line of code that reports the count of valid values in the last column (HINT: count())

In [83]:
df_temps[df_temps.columns[-1]].count()

0

### 9) Drop that last column from df_temps.
I cannot emphasize this enough – you will get the most out of your data when you take the time to set up the
most accurate type for each variable. Currently, the type of every variable is object. However, notice that in
your raw data file, EVERY variable is a number except the first variable, which is a date. Dates are COMMON in
data, and it is important that you represent dates as actual date types! We'll deal with that shortly. Let's
continue cleaning this up.

In [84]:
df_temps.drop(df_temps.columns[-1],axis = 1,inplace = True)

### 10) [P] Convert all numeric data to actual numeric data types. You'll need to look up how to do this. 
(HINT:
pd.to_numeric() is your friend.) Leave the NaN fields alone! The fact that they are missing is
IMPORTANT! And, leave the date/time variable in the first column alone.
You should output the shape of your data, and show info() to show every variable is a floating point
number except the date/time field in the first column, which should still be of type object.

In [85]:
for column in df_temps.columns[1:]:
    df_temps[column] = pd.to_numeric(df_temps[column])
print("shape: ",df_temps.shape)
df_temps.info()

shape:  (173252, 13)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173252 entries, 0 to 173251
Data columns (total 13 columns):
Date/Time (GMT)                  173252 non-null object
Number of Observations (n/a)     173252 non-null int64
Average Temp (F)                 172253 non-null float64
Max Temp (F)                     172253 non-null float64
Min Temp (F)                     172253 non-null float64
Average Dewpoint Temp (F)        172147 non-null float64
1 Hour Precip (in)               28934 non-null float64
Max Wind Gust (mph)              23544 non-null float64
Average Relative Humidity (%)    168754 non-null float64
Average Wind Speed (mph)         172710 non-null float64
Average Station Pressure (mb)    172974 non-null float64
Average Wind Direction (deg)     140590 non-null float64
Max Wind Speed (mph)             172710 non-null float64
dtypes: float64(11), int64(1), object(1)
memory usage: 17.2+ MB


### 11) [P] How much did our memory footprint improve? (Show the total memory usage using deep=True). Reportthe total memory usage in MB, and report the percentage improvement.

In [86]:
curMem = df_temps.memory_usage(deep = True).sum() / 1000000
print("new memory usage (MB): ",curMem)
print("new memory footprint is {:.2f}% of original".format(curMem/original_memory*100))

new memory usage (MB):  29.799472
new memory footprint is 20.96% of original


### 12) [P] Did you notice that to_numeric() has a parameter called downcast? Go back and read about this parameter. By default, most of the time your integer types will be converted to a 64-bit integer, and floating point types will use double precision numbers. You can do far better. Downcast your types accordingly. Report your latest memory usage in MB.

In [87]:
#i assumed we were supposed to downcast in the first time because all numbers needed to be floats, and without downcasting the "number of observations" was an int
#I will go back and change it. I think the question was unclear though
df_temps[df_temps.columns[1]] = pd.to_numeric(df_temps[df_temps.columns[1]],downcast = "integer")
for column in df_temps.columns[2:]:
    df_temps[column] = pd.to_numeric(df_temps[column],downcast = "float")
curMem = df_temps.memory_usage(deep = True).sum() / 1000000
print("new memory usage (MB): ",curMem)

new memory usage (MB):  20.96362


### 13) At this point, with the exception of the date column, you should have good data to start working with. Verify it by outputting the results of describe(). Every variable should have its basic stats reported!


In [88]:
df_temps.describe()

Number of Observations (n/a)  Average Temp (F)   Max Temp (F)  \
count                 173252.000000     172253.000000  172253.000000   
mean                       1.339915         51.286217      51.399208   
std                        0.854852         18.888988      18.907930   
min                        0.000000        -11.900000     -11.900000   
25%                        1.000000         36.000000      36.000000   
50%                        1.000000         52.000000      52.000000   
75%                        1.000000         66.900002      66.900002   
max                       10.000000        102.000000     102.000000   

        Min Temp (F)  Average Dewpoint Temp (F)  1 Hour Precip (in)  \
count  172253.000000              172147.000000        28934.000000   
mean       51.179626                  40.253628            0.030566   
std        18.881851                  19.053423            0.079247   
min       -11.900000                 -20.900000            0.000000   
25%        36.000000                  25.830000            0.000000   
50%        51.799999                  41.000000            0.000000   
75%        66.900002                  57.000000            0.030000   
max       102.000000                  79.000000            2.350000   

       Max Wind Gust (mph)  Average Relative Humidity (%)  \
count         23544.000000                  168754.000000   
mean             22.295341                      68.819511   
std               7.581788                      19.694031   
min               0.000000                       0.000000   
25%              19.600000                      54.000000   
50%              21.900000                      71.000000   
75%              26.500000                      86.000000   
max              88.599998                     100.000000   

       Average Wind Speed (mph)  Average Station Pressure (mb)  \
count             172710.000000                  172974.000000   
mean                   5.904806                    1016.741272   
std                    5.192565                       7.640043   
min                    0.000000                     508.600006   
25%                    0.000000                    1012.200012   
50%                    5.370000                    1016.900024   
75%                    9.200000                    1021.700012   
max                   76.000000                    1044.400024   

       Average Wind Direction (deg)  Max Wind Speed (mph)  
count                 140590.000000         172710.000000  
mean                     176.806366              6.175715  
std                      118.790619              5.311181  
min                        0.000000              0.000000  
25%                       70.000000              0.000000  
50%                      220.000000              5.800000  
75%                      280.000000              9.200000  
max                      360.000000             76.000000

### 14) [M] There are four primary classes in pandas for working with dates and times? Consider the Scalar Class for each, and state what concept each is representing.

 Date time: Timestamp - date and time and maybe also timezone  
Time delta: Timedelta - amount of time   
Time Span: Period - time metric - like a day or month or year. used often for indexing. eg period: january 2000  
Date offset: DateOffset  - like time delta but follows rules of calendar.accounts for leap years, daylight savings

### 15) [M] For each above, state the primary creation method used to create each type of data

Date time: to_datetime , date_range  
Time delta: to_timedelta , timedelta_range  
Time span: Period , period_range  
Date offset: DateOffset

### 16) [P] Create a Timestamp object from the string "07/04/19", which is a date representing July 4, 2019.
Store the object as d1 and show it.

In [89]:
d1 = pd.to_datetime("07/04/19")
d1

Timestamp('2019-07-04 00:00:00')

### 17) [P] Using d1 and string formatting codes, print the string from d1:
"Today's date is Thursday, July 4, 2019".

In [90]:
print("Today's date is {}, {} {}, {}".format(d1.day_name(),d1.month_name(),d1.day,d1.year))

Today's date is Thursday, July 4, 2019


### 18) [P] Create another Timestamp object representing Sept 7, 2019 at 3pm, called d2. Report it

In [91]:
d2 = pd.to_datetime("09/07/19, 3pm")
d2

Timestamp('2019-09-07 15:00:00')

### 19) [P] Subtract d2 – d1, and report the difference as the number of days and seconds between these two. Also report the difference as total seconds. (NOTE: The difference should be 65 days, 54000 seconds. Or 5670000 total seconds.)

In [92]:
diff = d2-d1
print("{} days and {} seconds".format(diff.days,diff.seconds))
print("{} seconds".format(diff.total_seconds()))

65 days and 54000 seconds
5670000.0 seconds


### 20) [P] Create a new Timestamp object from the string "2019-07-01 08:30pm", but, localize the time stamp to represent the time in the US Eastern Time Zone. Store the result as d3 and output it.

In [93]:
d3 = pd.to_datetime("2019-07-01 08:30pm").tz_localize("US/Eastern")
d3

Timestamp('2019-07-01 20:30:00-0400', tz='US/Eastern')

### 21) [P] Show time represented by d3, but converted to the US / Pacific Time Zone. The time reported should be three hours earlier than EST shown in the previous question.

In [94]:
d3.tz_convert("US/Pacific")

Timestamp('2019-07-01 17:30:00-0700', tz='US/Pacific')

### 22) [P] Create a Timestamp object representing right now, stored as ts_now. Show the result.

In [95]:
ts_now = pd.Timestamp.now()
ts_now

Timestamp('2020-02-02 19:44:04.603610')

### 23) [P] Create a Timedelta object representing 1 hour, stored as td_hour. Show the result.

In [96]:
td_hour = pd.to_timedelta("1h")
td_hour

Timedelta('0 days 01:00:00')

### 24) [P] Demonstrate how you can do basic mathematical operations by adding 6 hours to ts_now using td_hour and basic math operations. (i.e. No loops or further calculations necessary!)

In [97]:
ts_now + td_hour * 6

Timestamp('2020-02-03 01:44:04.603610')

### 25) [P] Create a DatetimeIndex object that represents every hour during the month of January, 2019. 
The first index should be midnight, January 1, 2020, and the last index should be January 31, 2020 at 11pm. Store the
object as dr. (HINT – use the pd.date_range() method!)

In [98]:
dr = pd.date_range("1/1/2020 12am","1/31/2020 11pm")

### 26) [P] The first variable in our data is currently an object. But, notice the name and its units? It's a date/time in the GMT time zone! Convert the first column of data into an actual time stamp.
NOTE: You can NOT simply generate this column using your own date range object! You must generate it
directly from the actual time/date stamp in the data! Why? This is very important. Do NOT ever be fooled
into thinking any real-world dataset you are dealing with is 100% complete. There are missing observations
in these data, and your data will be massively flawed if you neglect this! If you simply try to use a date range
between 1/1 – 12/31, with every hour, you are making an incorrect assumption that every observation is
present.
(HINT: Go back to your reference table. You are creating an array of timestamps. Which function? Either
to_datetime or date_range. We already told you that date_range is wrong!)

In [99]:
df_temps[df_temps.columns[0]] = pd.to_datetime(df_temps[df_temps.columns[0]]).dt.tz_localize("GMT")

### 27) [P] Confirm that your first column data type is now a timestamp by showing the output of df_temps.info(). (It should show that it is datetime64, to be exact) . Then, show the values of the first column of the first AND last row only. Your result should look like:
0 2000-01-01 00:00:00  
173251 2019-12-31 23:00:00

In [125]:
df_temps.info()
print(df_temps.iat[0,0])
print(df_temps.iat[-1,0])
# Should my result show index number? I don't know id I was supposed to do this differently or just get the index separately

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173252 entries, 0 to 173251
Data columns (total 13 columns):
Date/Time (GMT)                  173252 non-null datetime64[ns, GMT]
Number of Observations (n/a)     173252 non-null int8
Average Temp (F)                 172253 non-null float32
Max Temp (F)                     172253 non-null float32
Min Temp (F)                     172253 non-null float32
Average Dewpoint Temp (F)        172147 non-null float32
1 Hour Precip (in)               28934 non-null float32
Max Wind Gust (mph)              23544 non-null float32
Average Relative Humidity (%)    168754 non-null float32
Average Wind Speed (mph)         172710 non-null float32
Average Station Pressure (mb)    172974 non-null float32
Average Wind Direction (deg)     140590 non-null float32
Max Wind Speed (mph)             172710 non-null float32
dtypes: datetime64[ns, GMT](1), float32(11), int8(1)
memory usage: 8.8 MB
2000-01-01 00:00:00+00:00
2019-12-31 23:00:00+00:00


### 28) Finally, let's move that first column to be the new index for your dataframe. Use the set_index method of of df_temps to be the first column of data, then use the drop method to eliminate the first column. It is now your index, and thus there is no need to keep this information twice.

In [132]:
df_temps = df_temps.set_index(df_temps[df_temps.columns[0]])
df_temps = df_temps.drop(df_temps.columns[0], axis = 1)
df_temps.head()

Number of Observations (n/a)  Average Temp (F)  \
Date/Time (GMT)                                                             
2000-01-01 00:00:00+00:00                             1              26.1   
2000-01-01 01:00:00+00:00                             1              26.1   
2000-01-01 02:00:00+00:00                             1              26.1   
2000-01-01 03:00:00+00:00                             1              26.1   
2000-01-01 04:00:00+00:00                             1              26.1   

                           Max Temp (F)  Min Temp (F)  \
Date/Time (GMT)                                         
2000-01-01 00:00:00+00:00          26.1          26.1   
2000-01-01 01:00:00+00:00          26.1          26.1   
2000-01-01 02:00:00+00:00          26.1          26.1   
2000-01-01 03:00:00+00:00          26.1          26.1   
2000-01-01 04:00:00+00:00          26.1          26.1   

                           Average Dewpoint Temp (F)  1 Hour Precip (in)  \
Date/Time (GMT)                                                            
2000-01-01 00:00:00+00:00                       14.0                 NaN   
2000-01-01 01:00:00+00:00                       14.0                 NaN   
2000-01-01 02:00:00+00:00                       15.1                 NaN   
2000-01-01 03:00:00+00:00                       12.0                 NaN   
2000-01-01 04:00:00+00:00                       14.0                 NaN   

                           Max Wind Gust (mph)  Average Relative Humidity (%)  \
Date/Time (GMT)                                                                 
2000-01-01 00:00:00+00:00            20.700001                           59.0   
2000-01-01 01:00:00+00:00                  NaN                           59.0   
2000-01-01 02:00:00+00:00                  NaN                           62.0   
2000-01-01 03:00:00+00:00                  NaN                           54.0   
2000-01-01 04:00:00+00:00                  NaN                           59.0   

                           Average Wind Speed (mph)  \
Date/Time (GMT)                                       
2000-01-01 00:00:00+00:00                 17.299999   
2000-01-01 01:00:00+00:00                 16.100000   
2000-01-01 02:00:00+00:00                 15.000000   
2000-01-01 03:00:00+00:00                 16.100000   
2000-01-01 04:00:00+00:00                 12.700000   

                           Average Station Pressure (mb)  \
Date/Time (GMT)                                            
2000-01-01 00:00:00+00:00                    1015.200012   
2000-01-01 01:00:00+00:00                    1015.900024   
2000-01-01 02:00:00+00:00                    1016.299988   
2000-01-01 03:00:00+00:00                    1016.599976   
2000-01-01 04:00:00+00:00                    1017.299988   

                           Average Wind Direction (deg)  Max Wind Speed (mph)  
Date/Time (GMT)                                                                
2000-01-01 00:00:00+00:00                         280.0             17.299999  
2000-01-01 01:00:00+00:00                         280.0             16.100000  
2000-01-01 02:00:00+00:00                         280.0             15.000000  
2000-01-01 03:00:00+00:00                         270.0             16.100000  
2000-01-01 04:00:00+00:00                         280.0             12.700000

### 29) [P] Give one final report on the total memory usage, and also show the % memory reduction made compared to when you first loaded the data.
Again, please take this seriously. This is a substantial amount of memory saved! Why? Because you took the
time to properly process every column to have it represent its most accurate type, using the smallest type
necessary. HUGE savings!

In [133]:
curMem = df_temps.memory_usage(deep = True).sum() / 1000000
print("new memory usage (MB): ",curMem)
pct = curMem/original_memory*100
print("new memory footprint is {:.2f}% of original".format(pct))
print("that's a {:.2f}% reduction in size!".format(100-pct))

new memory usage (MB):  9.182356
new memory footprint is 6.46% of original
that's a 93.54% reduction in size!


### 30) [P] This dataset has missing observations. But, how many? First, calculate how many observations SHOULD be there. Use the difference between the first and last index value to compute this.

In [156]:
# ref : https://stackoverflow.com/questions/31283001/get-total-number-of-hours-from-a-pandas-timedelta
should = ((df_temps.index[-1] - df_temps.index[0]) / pd.Timedelta(hours=1) ) +1 # +1 b/c we want to include the start time
missing = int(should - len(df_temps))
print("there are {} missing observations".format(missing))

there are 2068 missing observations


### 31) [P] There are quite a lot! It's time to investigate. Create a data frame called df_missing that has an index of the time stamp of every missing date, with a simple variable called "missing" that has a value of 1 for every entry. (i.e. it should only contain the missing dates.) Report the number of rows in df_missing. It should match the number you computed previously.

In [157]:
dateRange = pd.date_range(df_temps.index[0],df_temps.index[-1],tz = "GMT",freq = "H")
#df_missing =
missingDates = dateRange.difference(df_temps.index)
# i am very proud i figured this out i had no idea how to do it at first
df_missing = pd.DataFrame({"missing":1},index = missingDates)
len(df_missing)

2068

### 32) [P] Let's get a sense of which years seem to be missing the most data. 

How? Well, the easiest approach is
probably to use the resample() method of data frames. Check out this section:
https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#resampling . This method works
phenomenally well for grouping and aggregating your data when you have a datetime index type!
We're going to resample our data by year, and perform a count aggregation all in one line:
Enter the following:
df_missing_by_year = df_missing.resample('Y').count()
There are many, many ways you can resample your data. You need to jump over to the options for dateoffset
objects. The letter codes are specified there:
https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#dateoffset-objects
Show the result of df_missing_by_year

In [159]:
df_missing_by_year = df_missing.resample('Y').count()
df_missing_by_year

missing
2000-12-31 00:00:00+00:00      792
2001-12-31 00:00:00+00:00       54
2002-12-31 00:00:00+00:00       30
2003-12-31 00:00:00+00:00       39
2004-12-31 00:00:00+00:00       72
2005-12-31 00:00:00+00:00      119
2006-12-31 00:00:00+00:00       32
2007-12-31 00:00:00+00:00       64
2008-12-31 00:00:00+00:00      193
2009-12-31 00:00:00+00:00       82
2010-12-31 00:00:00+00:00       40
2011-12-31 00:00:00+00:00       32
2012-12-31 00:00:00+00:00      152
2013-12-31 00:00:00+00:00       45
2014-12-31 00:00:00+00:00       38
2015-12-31 00:00:00+00:00       37
2016-12-31 00:00:00+00:00       27
2017-12-31 00:00:00+00:00      119
2018-12-31 00:00:00+00:00       67
2019-12-31 00:00:00+00:00       34

### 33) [P] You can see that pretty much every year has missing data. Not uncommon. However, one year in particular is really bad. Which one? Write the code to eliminate that entire year from df_year.
Congratulations! At this point, you performed your first real-world example of what you need to go through to
complete basic preprocessing steps!

In [170]:
#get year with most missing
yearToEliminate = df_missing_by_year["missing"].idxmax().year
# get indices of all rows in that year
indicesToEliminate = df_temps[df_temps.index.year == yearToEliminate].index
df_temps.drop(indicesToEliminate, inplace = True)
df_temps.head()

Number of Observations (n/a)  Average Temp (F)  \
Date/Time (GMT)                                                             
2001-01-01 00:00:00+00:00                             1              21.0   
2001-01-01 01:00:00+00:00                             1              19.9   
2001-01-01 02:00:00+00:00                             1              19.0   
2001-01-01 03:00:00+00:00                             1              19.0   
2001-01-01 04:00:00+00:00                             1              19.0   

                           Max Temp (F)  Min Temp (F)  \
Date/Time (GMT)                                         
2001-01-01 00:00:00+00:00          21.0          21.0   
2001-01-01 01:00:00+00:00          19.9          19.9   
2001-01-01 02:00:00+00:00          19.0          19.0   
2001-01-01 03:00:00+00:00          19.0          19.0   
2001-01-01 04:00:00+00:00          19.0          19.0   

                           Average Dewpoint Temp (F)  1 Hour Precip (in)  \
Date/Time (GMT)                                                            
2001-01-01 00:00:00+00:00                        6.1                 NaN   
2001-01-01 01:00:00+00:00                        7.0                 NaN   
2001-01-01 02:00:00+00:00                        7.0                 NaN   
2001-01-01 03:00:00+00:00                        7.0                 NaN   
2001-01-01 04:00:00+00:00                        6.1                 NaN   

                           Max Wind Gust (mph)  Average Relative Humidity (%)  \
Date/Time (GMT)                                                                 
2001-01-01 00:00:00+00:00                  NaN                           51.0   
2001-01-01 01:00:00+00:00                  NaN                           56.0   
2001-01-01 02:00:00+00:00                  NaN                           58.0   
2001-01-01 03:00:00+00:00                  NaN                           58.0   
2001-01-01 04:00:00+00:00                  NaN                           56.0   

                           Average Wind Speed (mph)  \
Date/Time (GMT)                                       
2001-01-01 00:00:00+00:00                      12.7   
2001-01-01 01:00:00+00:00                      10.4   
2001-01-01 02:00:00+00:00                      16.1   
2001-01-01 03:00:00+00:00                      15.0   
2001-01-01 04:00:00+00:00                      16.1   

                           Average Station Pressure (mb)  \
Date/Time (GMT)                                            
2001-01-01 00:00:00+00:00                    1019.000000   
2001-01-01 01:00:00+00:00                    1019.000000   
2001-01-01 02:00:00+00:00                    1019.000000   
2001-01-01 03:00:00+00:00                    1018.599976   
2001-01-01 04:00:00+00:00                    1018.599976   

                           Average Wind Direction (deg)  Max Wind Speed (mph)  
Date/Time (GMT)                                                                
2001-01-01 00:00:00+00:00                         270.0                  12.7  
2001-01-01 01:00:00+00:00                         270.0                  10.4  
2001-01-01 02:00:00+00:00                         270.0                  16.1  
2001-01-01 03:00:00+00:00                         260.0                  15.0  
2001-01-01 04:00:00+00:00                         280.0                  16.1

ref: https://thispointer.com/python-pandas-how-to-drop-rows-in-dataframe-by-conditions-on-column-values/